### PR TITLE
Pre-build python apks for slow compiling packages

### PR DESCRIPTION
**Pull request by ashb**
The APKBUILD (and patches) were taken from the Alpine edge repos. The
only reason we need to rebuild them is so that they work with _our_
version of python (which is 3.7 as of alpine:3.10, Alpine edge is now on
3.8).

The exact commits/locations of the APKBUILD files included in this
commit:

- https://git.alpinelinux.org/aports/tree/main/cython/APKBUILD?id=646e3a59d7dfc38e3527e3f9e9518e8fc02d3661
- https://git.alpinelinux.org/aports/tree/main/py3-bcrypt/APKBUILD?id=a6733d86cf6e3446e496eb56dc1a661475098eb1
- https://git.alpinelinux.org/aports/tree/testing/py3-cassandra-driver/APKBUILD?id=7d989d745a09fd352495857f98c6b1d2518b9d22
- https://git.alpinelinux.org/aports/tree/main/py3-cryptography/APKBUILD?id=af3d5861ca9c2000eb6fdb664919b509e4e7c136
- https://git.alpinelinux.org/aports/tree/community/py3-pycryptodome/APKBUILD?id=dfd7dd095f1736998ef02119cda68ec77f123e71
- https://git.alpinelinux.org/aports/tree/testing/py3-fastavro/APKBUILD?id=06e72c68723c0637a9d856b4b869c179e584bf3a
- https://git.alpinelinux.org/aports/tree/testing/py3-grpcio/APKBUILD?id=0f73e4c7e79be62b20c4e5eb8eb7d3473ecd934b
- https://git.alpinelinux.org/aports/tree/community/py3-numpy/APKBUILD?id=1d6fbc11289f873ae1d9903b9d673a59eec7e92c
- https://git.alpinelinux.org/aports/tree/testing/py3-pandas/APKBUILD?id=43f79fbbe1a0e4975da7a2649c7a3214d7d0396a
- https://git.alpinelinux.org/aports/tree/community/py3-pycryptodome/APKBUILD?id=dfd7dd095f1736998ef02119cda68ec77f123e71
- https://git.alpinelinux.org/aports/tree/main/py3-pynacl/APKBUILD?id=4c4f0346a603a43b44ebf2622a0cd881fa0427c2
- https://git.alpinelinux.org/aports/tree/testing/py3-typed-ast/APKBUILD?id=c240d3f1ca8dfe9d6af5d6a736da39b9a58ba6aa

We don't store the private key anywhere, so each time we need to build
new packages we will generate a new keypair. (This is fine, as we don't
leave the image with our repo configured, so we don't ever push out
updates that we expect `apk update` to pick up.)

I have used GIT LFS (Large File Storage) for the blobs so that a `git clone` is not so weighty. This may need `git lfs install` running locally if you want to update any of these files.

To build (or rebuild) the packages `make` from inside the `apk` folder should be enough.

### CODE REVIEW

File: docker/airflow/1.10.5/apks/apk-builder/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Based off https://github.com/andyshinn/docker-alpine-abuild but for apline 3.5
+FROM alpine:3.10
+RUN apk --no-cache add alpine-sdk coreutils \
+  && adduser -G abuild -g "Alpine Package Builder" -s /bin/ash -D builder \
```
Comment (schnie) [inline docker/airflow/1.10.5/apks/apk-builder/Dockerfile]:
```suggestion
  && adduser -G abuild -g "Alpine Package Builder" -s /bin/sh -D builder \
```
Should this be `/bin/sh`?

File: docker/airflow/1.10.5/apks/py3-numpy/numpy-1.17.0-musl.patch [hunk 2]
```diff
@@ -0,0 +1,39 @@
+diff --git a/numpy/core/include/numpy/npy_endian.h b/numpy/core/include/numpy/npy_endian.h
```
Comment (schnie) [inline docker/airflow/1.10.5/apks/py3-numpy/numpy-1.17.0-musl.patch]: @ashb what is this file for?

File: docker/airflow/1.10.5/Dockerfile [hunk 3]
```diff
@@ -27,6 +27,7 @@ LABEL io.astronomer.docker.airflow.version="1.10.5"
 ARG ORG="astronomer"
 ARG VERSION="1.10.5-2"
 ARG SUBMODULES="all, statsd, elasticsearch"
+ARG REPO_BRANCH=master
```
Comment (schnie) [inline docker/airflow/1.10.5/Dockerfile]: @ashb it feels a little weird to be always pulling artifacts from our `master` branch. Like if we're building from a release branch, we'd want to install from that branch probably right?

Probably not the biggest deal, just wanted to discuss.

Comment (ashb) [reply docker/airflow/1.10.5/Dockerfile]: Yes, this was the change I wanted to make after #514 was made.

File: docker/airflow/1.10.5/apks/py3-numpy/numpy-1.17.0-musl.patch (hunk 2 above)
Comment (ashb) [reply docker/airflow/1.10.5/apks/py3-numpy/numpy-1.17.0-musl.patch]: I honestly don't know - it was the patch that Alpine apply when they build their py3-numpy@edge package that we were installing before.

**Comment by ashb** [comment]
> this definitely bumps up the maintenance factor on this image.

It does, but the "failure" mode here is that we'll just fallback to compiling the package when we do `pip3 install --no-cache-dir "${AIRFLOW_MODULE}"` so that doesn't worry me.

> What's the process to update these packages?

Two ways of doing it: (which I'll add to a readme in the apks folder)

1. Get the latest build file from Alpine directly:
    1. Go to pkgs.alpinelinux.org and search for the package we want, for example `*bcrypt*`.
    2. Click on the package name
    3. Click on "Git repository" and replace the APKBUILD in our folder with the latest one.
2. To update a version manually, edit the APKBUILD file:
    1. Set `pkgver` to the latest relese
    2. Set `pkgrel` back to 0
    3. Update the `sha512sums` line, for example: `curl -fsSL https://github.com/pyca/bcrypt/archive/3.1.7.tar.gz | shasum -a 512`


Then `make -C docker/airflow/1.10.5/apks` should take care of the rest.

> Is it worth just doing it for all our packages to standardize?

I'd probably say not for anything that is quick enough to install via `pip`.

> Is there anything else about this worries you?

Not really - cos this just "pre-installs" versions, so if they version it installs isn't good enough `pip` will uninstall it and compile the new one.

File: docker/airflow/1.10.5/Dockerfile (hunk 3 above)
Comment (ashb) [reply docker/airflow/1.10.5/Dockerfile]: Now done via `bin/build-airflow`

File: docker/airflow/1.10.5/apks/apk-builder/Dockerfile (hunk 1 above)
Comment (ashb) [reply docker/airflow/1.10.5/apks/apk-builder/Dockerfile]: Could be, they are all symlinks in alpine to /bin/busybox

Comment (schnie) [reply docker/airflow/1.10.5/apks/apk-builder/Dockerfile]: oh, didn't realize alpine had that symlinked.

**Comment by schnie** [comment]
Tried running `make build-airflow` locally and got this:

```
Step 23/34 : COPY apks/keys/humans@astronomer.io.rsa.pub /etc/apk/keys
 ---> 8524d3761030
Step 24/34 : ENV PIP_CONSTRAINT=/usr/local/share/astronomer-pip-constraints.txt
 ---> Running in 5b31bf72ed6d
Removing intermediate container 5b31bf72ed6d
 ---> d4651a986241
Step 25/34 : COPY include/pip-constraints.txt /usr/local/share/astronomer-pip-constraints.txt
 ---> 1dc81585c692
Step 26/34 : RUN echo https://github.com/astronomer/astronomer/raw/${REPO_BRANCH}/docker/airflow/1.10.5/apks/repo >> /etc/apk/repositories      && apk update   && apk add --no-cache --virtual .build-deps         build-base               cyrus-sasl-dev          freetds-dev             freetype-dev            git             krb5-dev                libffi-dev              libxml2-dev             libxslt-dev             linux-headers                nodejs          nodejs-npm              python3-dev             tzdata  && apk add --no-cache           bash            ca-certificates                 cyrus-sasl              cython3     freetds          krb5-libs               py3-bcrypt              py3-cassandra-driver'>='3               py3-cryptography                py3-fastavro            py3-grpcio              py3-mysqlclient             py3-numpy-dev            py3-numpy               py3-pandas              py3-psycopg2            py3-pycryptodome                py3-pynacl              py3-typed-ast           python3                 tini    && update-ca-certificates    && cp /usr/share/zoneinfo/UTC /etc/localtime    && pip3 install --no-cache-dir --upgrade pip==19.3.1    && pip3 install --no-cache-dir --upgrade setuptools==41.0.1     && pip3 install --no-cache-dir --upgrade snowflake-connector-python==1.9.1   && pip3 install --no-cache-dir "${AIRFLOW_MODULE}"      && pip3 install --no-cache-dir "https://github.com/astronomer/astronomer-fab-securitymanager/releases/download/v1.2.1/astronomer_fab_security_manager-1.2.1-py3-none-any.whl"        && cd /usr/lib/python3.7/site-packages/airflow/www_rbac         && npm install  && npm run build        && rm -rf node_modules  && apk del .build-deps py3-numpy-dev         && ln -sf /usr/bin/python3 /usr/bin/python      && ln -sf /usr/bin/pip3 /usr/bin/pip    && sed -i -e '\#https://github.com/astronomer/astronomer/#d' /etc/apk/repositories
 ---> Running in f7edf6a58e10
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/community/x86_64/APKINDEX.tar.gz
fetch https://github.com/astronomer/astronomer/raw/master/docker/airflow/1.10.5/apks/repo/x86_64/APKINDEX.tar.gz
ERROR: https://github.com/astronomer/astronomer/raw/master/docker/airflow/1.10.5/apks/repo: No such file or directory
WARNING: Ignoring APKINDEX.4f8d869b.tar.gz: No such file or directory
v3.10.3-35-g855d7eeff4 [http://dl-cdn.alpinelinux.org/alpine/v3.10/main]
v3.10.3-24-gad8a2e9943 [http://dl-cdn.alpinelinux.org/alpine/v3.10/community]
1 errors; 10339 distinct packages available
The command '/bin/sh -c echo https://github.com/astronomer/astronomer/raw/${REPO_BRANCH}/docker/airflow/1.10.5/apks/repo >> /etc/apk/repositories       && apk update   && apk add --no-cache --virtual .build-deps build-base               cyrus-sasl-dev          freetds-dev             freetype-dev            git             krb5-dev                libffi-dev              libxml2-dev             libxslt-dev             linux-headers                nodejs          nodejs-npm              python3-dev             tzdata  && apk add --no-cache           bash            ca-certificates                 cyrus-sasl              cython3     freetds          krb5-libs               py3-bcrypt              py3-cassandra-driver'>='3               py3-cryptography                py3-fastavro            py3-grpcio              py3-mysqlclient             py3-numpy-dev            py3-numpy               py3-pandas              py3-psycopg2            py3-pycryptodome                py3-pynacl              py3-typed-ast           python3                 tini    && update-ca-certificates    && cp /usr/share/zoneinfo/UTC /etc/localtime    && pip3 install --no-cache-dir --upgrade pip==19.3.1    && pip3 install --no-cache-dir --upgrade setuptools==41.0.1     && pip3 install --no-cache-dir --upgrade snowflake-connector-python==1.9.1   && pip3 install --no-cache-dir "${AIRFLOW_MODULE}"      && pip3 install --no-cache-dir "https://github.com/astronomer/astronomer-fab-securitymanager/releases/download/v1.2.1/astronomer_fab_security_manager-1.2.1-py3-none-any.whl"        && cd /usr/lib/python3.7/site-packages/airflow/www_rbac         && npm install  && npm run build        && rm -rf node_modules  && apk del .build-deps py3-numpy-dev         && ln -sf /usr/bin/python3 /usr/bin/python      && ln -sf /usr/bin/pip3 /usr/bin/pip    && sed -i -e '\#https://github.com/astronomer/astronomer/#d' /etc/apk/repositories' returned a non-zero code: 1
make[1]: *** [Makefile:83: build-airflow] Error 1
make[1]: Leaving directory '/home/schnie/repos/astronomer'
make: *** [Makefile:33: build] Error 2
```

**Comment by schnie** [comment]
Oh, I guess because it wants to build them from `master` on github and they don't exist there yet.

**Comment by schnie** [comment]
@ashb is it possible to build using local `apk`s instead of pulling from github?

**Comment by ashb** [comment]
> @ashb is it possible to build using local `apk`s instead of pulling from github?

Hmmmm, not easily, as it needs to be pointed at an HTTP server.

`make build-airflow DRONE_BRANCH=alpine-packages` should work (to build the packages from this PR)`

**Pull request merged**